### ISSUE TITLE
Freeze the gradient expression tree

### ISSUE THREAD

**Issue by martin-ueding** [HIPS/autograd#471]
I have a Hybrid Monte Carlo simulation where we need to compute the gradient over and over. Our particular function works such that the structure of the derivative does not change, just the numbers that we plug in.

As I understand autograd it will evaluate the the function once with a wrapper around the data to see which expression tree comes out, then it builds the derivative from that. I know that for our case that this is not going to change, so if I could “freeze” this gradient expression tree it would save a lot of time for us.

In the profile I see that about 52 % of the time is spend building that tree and 47 % on evaluating it.

![snakeviz_001_initial_all](https://user-images.githubusercontent.com/976924/52120365-59db6100-261c-11e9-8453-2b3fefa2c022.png)

Is it possible to build the tree only once and then just plug in different numbers?

**Comment by martin-ueding** [comment]
That profile was done with `cPython` and then visualized with Snakeviz, really easy to use.

I have tried the `autograd.misc.const_graph` function on the gradient that I got with `autograd.grad` and instantly got the amortized speed-up of about 2 that can be seen in the above profile. Now the same thing finishes in half the time:

![snakeviz_005_const_graph](https://user-images.githubusercontent.com/976924/52139363-3cc08580-2650-11e9-8df2-c8a8e43c53f7.png)

This already is a great improvement, thank you very much for the pointer (and implementing/drafting it at some point)!

As I did not succeed with Numba and Autograd for certain functions not being implemented, I have then tried JAX. I needed to replace `jax.numpy.random` with plain `numpy.random` but as that is not part of the things that are differentiated, it makes no difference.

Though I got it working after several such fixes, there were a lot of outputs of this line:

> warning: `<unknown>:0:0:` loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering

The execution time of my program has increased greatly, taking 40 seconds for 3 iterations compared to the 15 seconds for 20 iterations. From testing 1, 2 and 3 iterations with the `jax.grad` function, I estimate that each iterations takes around 10 seconds and there is like 10 seconds setup time. I then added `jax.jit` around the gradient function, but that did not change anything.

I presume that I am doing something wrong, but that would rather be an issue over at JAX and not here.

**Comment by austinhoag** [comment]
Using `const_graph` will result in an error if you have a conditional in the function (or function called within that function) whose gradient you are evaluating. 

```python
from autograd import jacobian
from autograd.misc import const_graph

def myfunc(theta):
    if theta < 0:
        return theta**3
    return theta**2
grad_myfunc = grad(myfunc)

grad_myfunc(1.0)
# returns 2.0 as expected

# freeze the trace
grad_myfunc_accel = const_graph(grad_myfunc)

grad_myfunc_accel(1.0)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/tracer.py in new_box(value, trace, node)
    117     try:
--> 118         return box_type_mappings[type(value)](value, trace, node)
    119     except KeyError:

KeyError: <class 'numpy.bool_'>

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-34-e650aece83e7> in <module>
      1 grad_myscalarfun_accel = const_graph(grad_myscalarfun)
----> 2 grad_myscalarfun_accel(1.0)

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/misc/tracers.py in _fun(*args)
     42     maybe_cached_unary_fun = const_graph_unary(unary_fun)
     43     @wraps(fun)
---> 44     def _fun(*args): return maybe_cached_unary_fun(args)
     45     return _fun
     46 

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/misc/tracers.py in maybe_cached_fun(x)
     30         else:
     31             start_node = ConstGraphNode.new_root()
---> 32             end_value, end_node = trace(start_node, _fun.pop(), x)
     33             if end_node is None:
     34                 raise Exception("Output is independent of input")

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/tracer.py in trace(start_node, fun, x)
      8     with trace_stack.new_trace() as t:
      9         start_box = new_box(x, t, start_node)
---> 10         end_box = fun(start_box)
     11         if isbox(end_box) and end_box._trace == start_box._trace:
     12             return end_box._value, end_box._node

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/misc/tracers.py in <lambda>(args)
     39 def const_graph(fun, *args, **kwargs):
     40     partial_fun = partial(fun, *args, **kwargs)
---> 41     unary_fun = lambda args: partial_fun(*args)
     42     maybe_cached_unary_fun = const_graph_unary(unary_fun)
     43     @wraps(fun)

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/wrap_util.py in nary_f(*args, **kwargs)
     18             else:
     19                 x = tuple(args[i] for i in argnum)
---> 20             return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)
     21         return nary_f
     22     return nary_operator

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/differential_operators.py in grad(fun, x)
     23     arguments as `fun`, but returns the gradient instead. The function `fun`
     24     should be scalar-valued. The gradient has the same type as the argument."""
---> 25     vjp, ans = _make_vjp(fun, x)
     26     if not vspace(ans).size == 1:
     27         raise TypeError("Grad only applies to real scalar-output functions. "

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/core.py in make_vjp(fun, x)
      8 def make_vjp(fun, x):
      9     start_node = VJPNode.new_root()
---> 10     end_value, end_node =  trace(start_node, fun, x)
     11     if end_node is None:
     12         def vjp(g): return vspace(x).zeros()

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/tracer.py in trace(start_node, fun, x)
      8     with trace_stack.new_trace() as t:
      9         start_box = new_box(x, t, start_node)
---> 10         end_box = fun(start_box)
     11         if isbox(end_box) and end_box._trace == start_box._trace:
     12             return end_box._value, end_box._node

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/wrap_util.py in unary_f(x)
     13                 else:
     14                     subargs = subvals(args, zip(argnum, x))
---> 15                 return fun(*subargs, **kwargs)
     16             if isinstance(argnum, int):
     17                 x = args[argnum]

<ipython-input-32-e7090ac9b606> in myscalarfun(theta)
      1 def myscalarfun(theta):
----> 2     if theta < 0:
      3         return theta **3
      4     return theta**2

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/numpy/numpy_boxes.py in __lt__(self, other)
     43     def __gt__(self, other): return anp.greater(self, other)
     44     def __ge__(self, other): return anp.greater_equal(self, other)
---> 45     def __lt__(self, other): return anp.less(self, other)
     46     def __le__(self, other): return anp.less_equal(self, other)
     47     def __abs__(self): return anp.abs(self)

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/tracer.py in f_wrapped(*args, **kwargs)
     39             argvals = subvals(args, [(argnum, box._value) for argnum, box in boxed_args])
     40             if f_wrapped in notrace_primitives[node_constructor]:
---> 41                 return f_wrapped(*argvals, **kwargs)
     42             parents = tuple(box._node for _     , box in boxed_args)
     43             argnums = tuple(argnum    for argnum, _   in boxed_args)

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/tracer.py in f_wrapped(*args, **kwargs)
     44             ans = f_wrapped(*argvals, **kwargs)
     45             node = node_constructor(ans, f_wrapped, argvals, kwargs, argnums, parents)
---> 46             return new_box(ans, trace, node)
     47         else:
     48             return f_raw(*args, **kwargs)

~/anaconda3/envs/seldo-test/lib/python3.8/site-packages/autograd/tracer.py in new_box(value, trace, node)
    118         return box_type_mappings[type(value)](value, trace, node)
    119     except KeyError:
--> 120         raise TypeError("Can't differentiate w.r.t. type {}".format(type(value)))
    121 
    122 box_types = Box.types

TypeError: Can't differentiate w.r.t. type <class 'numpy.bool_'>
```
Is there a workaround? I wonder if the reason this happens is related to the JAX sharp bit when using JIT and conditionals: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#python-control-flow-jit
